### DISCUSSION TITLE
Custom parametrized extractor

### DISCUSSION THREAD

**Discussion by czocher** [tokio-rs/axum#1533] [Q&A] [answered]
Is there some way to create an extractor which would take additional parameters? What I had in mind is some kind of extractor to get a user only if it has a certain role assigned, something like this:

```rust
fn some_handler(RequireRole(admin): RequireRole(Role::Admin)) -> IntoResponse {
    ...
}
```

where the `RequireRole` would use a different extractor to get a `user` and then verify if `user.role == Role::Admin`. I see that the `FromRequest` or `FromRequestParts` traits don't provide a `self` param which kinda limits their use in this manner, but maybe there's some way to overcome that, which I'm not aware of?

**Comment by jplatte** [comment, marked as answer]
This is possible on Nightly where you can declare a const-generic `RequireRole<const R: Role>` with the right unstable feature (I don't know the name off the top of my head but the compiler error message should mention it). It would then be used as `RequireRole(admin): RequireRole<Role::Admin>`.

On stable, you can declare a `RequireRole<R: Role>` with `Admin` being one of several zero-sized types that implement the `Role` trait. However, in that case you need to add a `PhantomData` field to `RequireRole` and so the pattern match will have to be `RequireRole(admin, _): RequireRole<Admin>` if you still want to use destructuring.

**Reply by czocher** [reply]
Thank you for confirming my conclusion. I understand that there are no other options for implementing this as an `extractor`? It's either one of your proposals or a `middleware`?

**Reply by jplatte** [reply]
Yes, fundamentally you can only pass arguments to extractors using generics. At runtime, an extractor only exists as a thin wrapper around the inner value, so it doesn't make sense for there to be a `self` argument in `FromRequest` – that `self` value only exists once the extractor has already done its work.
